### ISSUE TITLE
dm-related errors when booting default mkosi image

### ISSUE THREAD

**Issue by keszybz** [systemd/systemd#27682]
### systemd version the issue has been seen with

git

### CPU architectures issue was seen on

None

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
During boot:

May 18 15:49:17 fedora systemd-udevd[231]: erofs: Device is queued (SEQNUM=1760, ACTION=add)
May 18 15:49:17 fedora kernel: /dev/mapper/usr: Can't open blockdev
May 18 15:49:17 fedora kernel: erofs: (device dm-0): mounted with root inode @ nid 36.
May 18 15:49:17 fedora mount[331]: mount: /sysusr/usr: WARNING: source write-protected, mounted read-only.
May 18 15:49:17 fedora systemd[1]: Received SIGCHLD from PID 331 (mount).
May 18 15:49:17 fedora systemd[1]: Child 331 (mount) died (code=exited, status=0/SUCCESS)
May 18 15:49:17 fedora systemd[1]: sysusr-usr.mount: Child 331 belongs to sysusr-usr.mount.
May 18 15:49:17 fedora systemd[1]: sysusr-usr.mount: Mount process exited, code=exited, status=0/SUCCESS (success)
May 18 15:49:17 fedora systemd[1]: sysusr-usr.mount: Changed mounting-done -> mounted
May 18 15:49:17 fedora systemd[1]: sysusr-usr.mount: Job 90 sysusr-usr.mount/start finished, result=done
May 18 15:49:17 fedora systemd[1]: Mounted sysusr-usr.mount.
```
I'm not sure what the error means, because `/usr` is mounted successfully.

During poweroff:
```
[  OK  ] Stopped systemd-tmpfiles-setup.service.
[  OK  ] Stopped target local-fs.target.
         Unmounting run-credentials-\x40system.mount...
         Unmounting run-credentials…\x2dfirstboot.service.mount...
         Unmounting run-credentials…d\x2dsysusers.service.mount...
         Unmounting tmp.mount...
[  OK  ] Stopped systemd-pcrfs-root.service.
[  OK  ] Stopped systemd-pcrmachine.service.
[  820.756919] systemd-veritysetup[709]: Device usr is still in use.
[  820.757840] systemd-veritysetup[709]: Failed to deactivate: Device or resource busy
[  820.760115] systemd[1]: systemd-veritysetup@usr.service: Failed with result 'exit-code'.
[  OK  ] Stopped systemd-veritysetup@usr.service.
[  OK  ] Unmounted run-credentials-\x40system.mount.
[  OK  ] Unmounted run-credentials-…md\x2dfirstboot.service.mount.
[  OK  ] Unmounted run-credentials-…emd\x2dsysusers.service.mount.
[  OK  ] Unmounted tmp.mount.
[  OK  ] Removed slice system-systemd\x2dveritysetup.slice.
[  OK  ] Stopped target local-fs-pre.target.
[  OK  ] Stopped target swap.target.
[  OK  ] Reached target umount.target.
[  OK  ] Stopped systemd-tmpfiles-setup-dev.service.
[  OK  ] Stopped systemd-sysusers.service.
[  OK  ] Stopped systemd-firstboot.service.
[  OK  ] Stopped systemd-remount-fs.service.
[  OK  ] Reached target shutdown.target.
[  OK  ] Reached target final.target.
[  OK  ] Finished systemd-poweroff.service.
[  OK  ] Reached target poweroff.target.
[  820.874695] systemd-shutdown[1]: Could not detach DM /dev/dm-0: Device or resource busy
[  820.875206] systemd-shutdown[1]: Failed to finalize DM devices, ignoring.
[  820.877702] reboot: Power down
```
```

**Comment by poettering** [comment]
I don't grok the `kernel: /dev/mapper/usr: Can't open blockdev` message. I am seeing the same here though. The message is generated by fs/super.c in the kernel, but I don't grok why/when this would be generated.

The systemd-shutdown message is kinda expected, if the system runs off /usr/ then it will remain busy till the very end. Only an exitrd could detach that...

The veritysetup message suggest we need to some conditionalization there, i.e. something like x-initrd.mount= for a volume backing /usr/

**Comment by keszybz** [comment]
> `kernel: /dev/mapper/usr: Can't open blockdev`
> May 18 15:49:17 fedora mount[331]: mount: /sysusr/usr: WARNING: source write-protected, mounted read-only.

Maybe the first message is just about not being able to open read-write? Since the usr partition is erofs, we should open it ro explicitly.

**Comment by poettering** [comment]
So the main issue is addressed by #27889, the "Can't open blockdev" issue is removed by #27895. 

The systemd-shutdown one I am not sure we should try to address:

```
[  820.874695] systemd-shutdown[1]: Could not detach DM /dev/dm-0: Device or resource busy
[  820.875206] systemd-shutdown[1]: Failed to finalize DM devices, ignoring.
```

maybe we can reword this to tone it down a bit so that it becomes clear this isn't really a problem.

**Comment by DaanDeMeyer** [comment]
> I guess with all three PRs this should be thoroughly addressed (though one message remains, indicating that /usr/ is not detached at boot, which I think should remain, since it is relevant info, the right way to address that would be to make mkosi define an exitrd)

Doesn't that need work in systemd first? We don't support exitrds right?

**Comment by poettering** [comment]
> Doesn't that need work in systemd first? We don't support exitrds right?

Well, not as a separate DDI. But we do support that initrds place enough for shutting down in /run/initramfs/ and then we'll transition into that during shutdown and it's job is to destroy what is remaining.

often it might actually suffice putting a copy of systemd-shutdown there actually (plus libs)
